src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -0,0 +1,688 @@
+                return GetUsingStatementsString(usingStatements);
+            }
+
+            private static string GetUsingStatementsString(HashSet<string> usingStatements)
+            {
+                string[] usingsArr = usingStatements.ToArray();
+                Array.Sort(usingsArr);
+                return string.Join("\n", usingsArr);
+            }
+
+            private static HashSet<string> GetUsingStatements(TypeMetadata typeMetadata)
+            {
+                HashSet<string> usingStatements = new();
+
+                // Add library usings.
+                usingStatements.Add(FormatAsUsingStatement("System.Runtime.CompilerServices"));
+                usingStatements.Add(FormatAsUsingStatement("System.Text.Json"));
+                usingStatements.Add(FormatAsUsingStatement("System.Text.Json.Serialization"));
+                usingStatements.Add(FormatAsUsingStatement("System.Text.Json.Serialization.Metadata"));
+
+                // Add imports to root type.
+                usingStatements.Add(FormatAsUsingStatement(typeMetadata.Type.Namespace));
+
+                switch (typeMetadata.ClassType)
+                {
+                    case ClassType.Nullable:
+                        {
+                            AddUsingStatementsForType(typeMetadata.NullableUnderlyingTypeMetadata!);
+                        }
+                        break;
+                    case ClassType.Enumerable:
+                        {
+                            AddUsingStatementsForType(typeMetadata.CollectionValueTypeMetadata);
+                        }
+                        break;
+                    case ClassType.Dictionary:
+                        {
+                            AddUsingStatementsForType(typeMetadata.CollectionKeyTypeMetadata);
+                            AddUsingStatementsForType(typeMetadata.CollectionValueTypeMetadata);
+                        }
+                        break;
+                    case ClassType.Object:
+                        {
+                            if (typeMetadata.PropertiesMetadata != null)
+                            {
+                                foreach (PropertyMetadata metadata in typeMetadata.PropertiesMetadata)
+                                {
+                                    AddUsingStatementsForType(metadata.TypeMetadata);
+                                }
+                            }
+                        }
+                        break;
+                    default:
+                        break;
+                }
+
+                void AddUsingStatementsForType(TypeMetadata typeMetadata)
+                {
+                    usingStatements.Add(FormatAsUsingStatement(typeMetadata.Type.Namespace));
+
+                    if (typeMetadata.CollectionKeyTypeMetadata != null)
+                    {
+                        Debug.Assert(typeMetadata.CollectionValueTypeMetadata != null);
```

Comment:
Why the Debug.Assert on `CollectionValueTypeMetadata`?

Reply:
I just expect that if there's metadata for a key type (for dictionaries), there should also be metadata for the corresponding value type. Can remove the assert since issues with this will be self-evident.